lib/ex3/index.ts

```diff
@@ -0,0 +1,218 @@
+import crypto from "crypto";
+import * as crypto2 from "@node-lightning/crypto";
+import { BufferReader, BufferWriter } from "@node-lightning/bufio";
+
+/**
+ * This reads a simplified onion packet. Simplifications are:
+ * - Reuse of the same ephemeral key at each hop
+ * - Payload encrypted
+ * - Use of variable length onion (grows in length with each hop)
+ *
+ * The packet parsing algorithm extracts from the packet:
+ * - [1]   - version
+ * - [33]  - ephemeral point (compressed secp256k1)
+ * - [var] - variable length payload (Encrypted)
+ * - [32]  - HMAC
+ *
+ * From there, the payload is decrypted and then parsed:
+ * - [1]   - data length
+ * - [var] - data used by this hop
+ * - [var] - data sent to next hop
+ * - [32]  - HMAC covering the next packet constuction
+ *
+ * Once this information is parsed, we construct and return the next
+ * packet.
+ *
+ * @param packet
+ * @param nodeKeys
+ */
+export function read(packet: Buffer, nodeKeys: Buffer[]): Buffer {
+  console.log("packet  ", packet.toString("hex"));
+
+  const packetReader = new BufferReader(packet);
+
+  // Read the version, we should always expect 0x00
+  const version = packetReader.readUInt8();
+  console.log("packet version ", version);
+
+  // Read the ephemeral point, in this example we reuse the same point
+  // which would allow correlation of the onion.
+  const ephemeralPoint = packetReader.readBytes(33);
+  console.log("packet ep      ", ephemeralPoint.toString("hex"));
+
+  // Read the payload from the packet
+  const payload = packetReader.readBytes(packet.length - 33 - 32 - 1);
+  console.log("packet payload ", payload.toString("hex"));
+
+  // Read the HMAC which is the final 32 bytes
+  const hmac = packetReader.readBytes();
+  console.log("packet hmac    ", hmac.toString("hex"));
+
+  // Construct a shared secret using ECDH using the ephemeral point and
+  // our public key.
+  const nodeKey = nodeKeys.shift();
+  const sharedSecret = crypto2.ecdh(ephemeralPoint, nodeKey);
+
+  // Compute an HMAC using the shared secret and the packet (less the
+  // HMAC). We can't compute the HMAC using the full packet, because
+  // the HMAC can't include itself haha.
+  const calcedHmac = crypto2.hmac(
+    sharedSecret,
+    packet.slice(0, packet.length - 32)
+  );
+  console.log("processed hmac ", calcedHmac.toString("hex"));
+
+  // Fail the packet if the HMAC is not the expected value!
+  if (!crypto.timingSafeEqual(hmac, calcedHmac)) {
+    throw new Error("HMAC failed");
+  }
+
+  // Decrypt the payload using shared secret 
+  const decrypt_payload = crypto2.ccpDecrypt(
```

Comment:
Sorry I misspoke earlier about what to do for encryption/decryption.  Chacha20-Poly1305 is used as   an AEAD cipher and will add a 16-byte MAC.  Since we're already using HMAC over the packet we don't need the MAC or AEAD functionality.  In this case we can just use a  ChaCha20 stream cipher as in https://github.com/altangent/node-lightning/blob/eacb6034771df4891c78ded7a5f29c572d541c5a/packages/crypto/lib/chacha.ts#L3.  

More info on using ChaCha20 directly: https://www.derpturkey.com/chacha20poly1305-aead-with-node-js/


As an example, you can see it used during packet construction here:
- https://github.com/altangent/node-lightning/blob/eacb6034771df4891c78ded7a5f29c572d541c5a/packages/onion/lib/generateCipherStream.ts#L3
- https://github.com/altangent/node-lightning/blob/eacb6034771df4891c78ded7a5f29c572d541c5a/packages/onion/lib/OnionPacket.ts#L69


One note is that calling `.update(data)` or generating a stream from an empty buffer (as in `generateCipherStream`) and xor'ing against the data as in the `onion` package above have the same results.   Breaking it into two pieces for the example would probably make it confusing.  So we can probably stick with `chachaEncrypt/Decrypt` methods.